### ISSUE TITLE
Improve our job execution model

### ISSUE THREAD

**Issue by jdoliner** [pachyderm/pachyderm#905]
Right now job sharding is very static, we compute how many shards we want to run before we start the job and create a k8s job with that much parallelism. Each of these jobs gets handed a chunk of work when start up and contact pachd. If each of the pods doesn't run their chunk of work to completion you get weird failure modes. For example if a pod dies then it will get rescheduled by k8s but when it contacts pachd it will get an error because the job already has too many pods. Meanwhile the shard it was processing will never be reassigned so 

We can fix this with a lease based model, this would make it so that even if one of the pods fails pachd can reassign the chunk of work it was doing to another pod. This also allows us to do some smart optimizations like if a lease expires on a shard we can split it up into 2 smaller chunks and schedule those. In theory if our logic in there got smart enough users wouldn't have much reason to specify parallelism for their jobs and could just rely on the system to figure it out for them.

**Comment by derekchiang** [comment]
Closed by #1022
